### PR TITLE
sw - create save image link

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

_Assignees_: @HaominZhang, @XanderStrike

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [x] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

### PR DESCRIPTION
**Pull request by swangs**
Closes #2 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: db/migrate/20180612224718_create_images.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class CreateImages < ActiveRecord::Migration[5.1]
```
Comment (HaominZhang) [inline db/migrate/20180612224718_create_images.rb]: How do you think make migration as a separate commit? My reason is that it is pretty independent with other rails code change and if anything goes wrong, we can easily revert itself instead of revert other changes.

File: app/views/images/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<form action="/images" method="POST">
+
+  <%= hidden_field_tag :authenticity_token, form_authenticity_token -%>
+
+  <input type="text" name="image[url]">
+  <input type="submit" value="Submit">
+</form>
+
+<% if flash[:errors]%>
```
Comment (HaominZhang) [inline app/views/images/new.html.erb]: From a user experience level I think it is better to put the flash message at the top

File: app/views/images/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<form action="/images" method="POST">
```
Comment (HaominZhang) [inline app/views/images/new.html.erb]: There is a gem called `simple_form` which gives us more flexibility with form. You can go thru the documentation on their github page if you are interested, but you don't have to change the code

File: config/routes.rb [hunk 4]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
+
+  resources :images, only: %i[show new create]
```
Comment (HaominZhang) [inline config/routes.rb]: 👍

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  def new; end
+
+  def create
+    image = Image.new
+    image.url = image_params[:url]
+
+    if image.save
+      redirect_to action: 'show', id: image.id
```
Comment (HaominZhang) [inline app/controllers/images_controller.rb]: I think an easier way to do it is `redirect_to image` :)

File: app/models/image.rb [hunk 6]
```diff
@@ -1,10 +1,11 @@
 class Image < ApplicationRecord
-  validates :url, presence: true, uniqueness: true
+  validates :url, uniqueness: true
```
Comment (HaominZhang) [inline app/models/image.rb]: What's the reason of this change?

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -1,7 +1,23 @@
...
-  # test "the truth" do
-  #   assert true
-  # end
+  test 'should show new page' do
+    get new_image_path
+    assert_response :success
+  end
+
+  test 'should create an image' do
+    post '/images',
```
Comment (HaominZhang) [inline test/controllers/images_controller_test.rb]: can we use rails path here instead of the actually url?

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -1,7 +1,23 @@
...
+         params: {
+           image: {
+             url: 'https://d2pu2bk1b66iw6.cloudfront.net/photos/2014/08/01/6-74677-mm_babymeme3-1406927575.jpg'
+           }
+         }
+    assert_response :redirect
+  end
+
+  test 'should redirect on invalid id' do
+    get '/images/10'
```
Comment (HaominZhang) [inline test/controllers/images_controller_test.rb]: can we use rails path here instead of the actually url?

File: test/models/image_test.rb [hunk 9]
```diff
@@ -1,7 +1,26 @@
 require 'test_helper'
 
 class ImageTest < ActiveSupport::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+  test 'should require url' do
+    image = Image.new
+    assert_not image.save, 'Saved the image without a url'
```
Comment (HaominZhang) [inline test/models/image_test.rb]: Instead of doing a real save action, we can call `valid?` on that object. And using `assert_not_valid_on` assertion would be better

File: test/controllers/images_controller_test.rb [hunk 10]
```diff
@@ -1,7 +1,23 @@
 require 'test_helper'
 
 class ImagesControllerTest < ActionDispatch::IntegrationTest
```
Comment (HaominZhang) [inline test/controllers/images_controller_test.rb]: I think in general, make the test go with its implementation is better rather than put all tests in one commit. Do you want to make some changes?

File: app/models/image.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class Image < ApplicationRecord
```
Comment (HaominZhang) [inline app/models/image.rb]: I think creating the model could be in its own commit as well. How do you think?

File: app/models/image.rb (hunk 6 above)
Comment (swangs) [reply app/models/image.rb]: When I had the presence check, there would be two errors that show up. One for the presence: true failing and one for the custom validation failing. I merged presence check into the custom validation so that there is a single error if any condition fails. Uniqueness check will never double fail because if it already exists in the database, it has already passed all the tests.

**Comment by swangs** [comment]
Changed all of the suggestions. Please review again.
